src/vs/workbench/contrib/debug/browser/debugSession.ts

```diff
@@ -1245,6 +1245,30 @@ export class DebugSession implements IDebugSession {
 				this.cancelAllRequests();
 				this.model.clearThreads(this.getId(), true);
 				await this.fetchThreads(this.getStoppedDetails());
+
+				const focusedThread = this.debugService.getViewModel().focusedThread;
+				const focusedThreadDoesNotExist = focusedThread !== undefined && focusedThread.session === this && !this.threads.has(focusedThread.threadId);
+				if (focusedThreadDoesNotExist) {
+					this.debugService.focusStackFrame(undefined, undefined);
+				}
+				const thread = typeof event.body.threadId === 'number' ? this.getThread(event.body.threadId) : undefined;
+				if (thread) {
+					const promises = this.model.refreshTopOfCallstack(<Thread>thread);
+					const focus = async () => {
+						if (focusedThreadDoesNotExist || thread.getCallStack().length) {
+							const focusedStackFrame = this.debugService.getViewModel().focusedStackFrame;
+							if (!focusedStackFrame || focusedStackFrame.thread.session === this) {
+								// Only take focus if nothing is focused, or if the focus is already on the current session
+								const preserveFocus = !this.configurationService.getValue<IDebugConfiguration>('debug').focusEditorOnBreak;
+								await this.debugService.focusStackFrame(undefined, thread, undefined, { preserveFocus });
+							}
+						}
+					};
+					await promises.topCallStack;
+					focus();
+					await promises.wholeCallStack;
+					focus();
```

Comment:
Are you sure that calling focus() twice is right?

Reply:
The code in the stop event handler (https://github.com/microsoft/vscode/blob/41f18661e4af20b9ac04fb06f0d8c70ea072ac85/src/vs/workbench/contrib/debug/browser/debugSession.ts#L979) has a similar construct, and this seems to be used to keep the UI performant in the case of a large stack which may take a long-time to read from the debug adapter.
If the InvalidatedEvent is received for the 'stack' we don't know which part of the stack has changed so we need to read the entire stack again, and could also suffer from this issue?